### DISCUSSION TITLE
Python functions help

### DISCUSSION THREAD

**Discussion by TheNose1** [community/community#111026] [Programming Help]
### Body

Hello all, 
I have been working with python for a couple months now getting used to the language and how to solve problems. I however am struggling with defining and calling my own functions. I cannot follow others code due to this issue and I struggle to make more complex programs due to this issue. So I am currently taking CS50P to help with this issue but am still stuck, particualarly with a class assignment using functions. I have written the code I need and it accomplishes the desired task but without using functions as deisred. I have commented out the parts of the code I previously wrote and have left the code I know works. If anyone could assist me with this lack of knowledge that would be very helpful.
![Screenshot 2024-03-05 144157](https://github.com/community/community/assets/161993218/afed8e92-fe87-4790-b53f-4017be01838c)

I do not now how to attach code files I am sorry

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by ethanappgenius** [comment]
It's great that you're seeking assistance to improve your understanding of defining and calling functions in Python. To address your struggle, let's focus on defining functions and calling them appropriately. In your provided code, you're on the right track with the 'convert' function, which takes 'time' as an argument and converts it to hours and minutes. However, you're immediately reassigning 'time' with a user input, which negates the purpose of passing arguments to the function. Instead, you should remove the input prompt from within the 'convert' function and call it with a time argument from your main code. Additionally, uncomment the 'main' function and the conditional block at the bottom of your script to properly call and execute your code. This separation of concerns will improve readability and maintainability, allowing you to tackle more complex programs effectively. Keep practicing, and don't hesitate to seek further clarification or assistance when needed.

**Reply by TheNose1** [reply]
I appreciate your response and it helped tremendously. I have reworked the code to work as intended with the functions. 
![Screenshot 2024-03-05 155841](https://github.com/community/community/assets/161993218/65999af6-7ff3-4d6e-9d18-0ce54a96c094)
After compiling however I get the correct answer but with an added printout of the now 'act_time' before the if statement prints. 
I do not know why this is there and cannot see where this is coming from in the code. Also the code seems to run with or without 'convert(time)' being called at the bottom of the "def main()" function. Why is that?

**Reply by TheNose1** [reply]
![Screenshot 2024-03-05 155945](https://github.com/community/community/assets/161993218/736d0f7c-2358-454a-afa2-f86566efb8d1)
Sorry here is the result after compiling

**Comment by vineeth-au** [comment]
How about we start with something simple... try this out.

```
def main()
    print("Hello World")

def convert()
    print("This function converts stuff!!!')

# The below statement checks if the script is being run as the main program. If it is, it calls the main() function after that the convert function.
if __name__ == '__main__':
    main()
    convert()

```

Try running this and see what happens!!!

**Comment by AmirSarrafzadeh** [comment]
Dear TheNose1,

I hope this message finds you well,

![image](https://github.com/community/community/assets/81471063/b8f690d6-2df5-41fc-80a5-6c5b366645cf)

Hope helps

**Comment by ShenalSen** [comment]
Hi ..TheNose1,
It’s great that you’re learning Python and working on solving problems! Let’s dive into functions and how to define and call them.

First, let’s address the code snippet you provided. I see that you’ve commented out some parts and left the working code intact. Let’s break it down:

`# def convert(time):
#     hours = int(time.split(":")[0])
#     minutes = int(time.split(":")[1])
#     return (hours * 60 + minutes) / 60
# print(convert("9:13"))

def input("What time is it?"):
    if 5 <= time <= 8:
        print("Breakfast time")
    elif 12 <= time <= 13:
        print("Lunch time")
    elif 18 <= time <= 19:
        print("Dinner time")
`

The commented-out function convert seems to convert a time in the format “hh:mm” to hours. You split the input time string, convert hours and minutes to integers, and calculate the total hours. However, it’s not being used in the current code.
The function input (which should be renamed since input is a built-in Python function) takes user input for the current time. However, there are a couple of issues:
The function definition should have parentheses with parameters (e.g., def input_time(prompt):).
You’re using an undeclared variable time inside the function. You need to get the user input and assign it to time.
Now, let’s refactor this code to use functions correctly:

`def get_user_time():
    return input("What time is it? (hh:mm): ")

def check_meal_time(time):
    hours = int(time.split(":")[0])
    if 5 <= hours <= 8:
        print("Breakfast time")
    elif 12 <= hours <= 13:
        print("Lunch time")
    elif 18 <= hours <= 19:
        print("Dinner time")
    else:
        print("Not a mealtime")

user_time = get_user_time()
check_meal_time(user_time)
`
Here’s what we did:

Created a get_user_time function to get the user’s input.
Renamed the function to check_meal_time for clarity.
Used the correct variable (hours) inside the check_meal_time function.
Added an “else” statement to handle non-mealtime hours.
Feel free to modify and expand upon this example. Functions allow you to organize your code, reuse logic, and make it more readable. Keep practicing, and you’ll get the hang of it! 😊
